Peer-review of v1.0(#3) (#4)
smart-online-judge/basic-online-judge: README.md

Аспект пропозиції | Суть аспекту пропозиції
:---              | :---
Назва програмної системи | Fair Online Judge

> Призначення системи. Порівняння з існуючими аналогічними системами. Ефект від її впровадження | <p>Основною задачею розробленої системи (тут і далі наз. **ОЗ**), у противагу до прикладної її складової, яка використовує та будує інфраструктуру навколо **ОЗ**) є визначення близькості текстових файлів (зокрема програмних кодів) у структурному та семантичному розумінні, візуалізація означеної близькості.</p><p>Інфраструктурна складова проекту використоує розроблену технологію **ОЗ** у задачі автоматичного оцінювання близькості (у означеному вище розумінні) текстових даних (у випадку **І2** тзокрема програмних кодів) (у навчальному середовищі). Ми виділяємо дві підзадачі, впорядковані за пріоритетністю виконання:<ul><li>автоматичне оцінювання близькості текстових файлів загальної форми (завдання у з вільною відкритою відповіддю; тут і далі **І1**)</li><li>автоматичне оцінювання близькості програмних кодів як задач з наявністю машинно-перевірюваного критерію правильності (тут і далі **І2**)</li></ul></p><p>**І1** та **І2** мають бути вирішені і розгорнуті у вигляді веб-сервера.</p></p>На даний авторам не відомий сервіс, що виконує задачі **ОЗ**, **І1** та **І2** разом. Існує вирішення задач **І1** та **І2** у окремих сервісах (Google Classroom та Codeforces відповідно). Не існує також розповсюдженого рішення задачі **ОЗ**.</p><p>Задача **ОЗ** не вирішена в загальному вигляді та є дослідницьким напрямком, у якому state-of-the-art технології продовжуюсть з'являтися. Тому інтегрування таких технологій у прикладні рішення має цінність для стимулювання розвитку цієї галузі.</p>

Author: - Вместо того, чтобы писать юзкейс, написал общую формулировку задачи. Интеграция с Google Drive не существенна. Суть исходного текста сохранил. 

- По поводу юзкейса - в доке написано, как будет выглядеть процесс работы с системой со стороны пользователя, что это будет либо local-app, либо web-app. "Либо" значит, что это не продумано. Нужно выбрать что-то одно, и я предлагаю выбрать веб-приложение.
- Приоритезировал этапы разработки приложения. Тебе интересна **ОЗ**, мне интересно построить инфраструктуру вокруг нее, причем в формулировке **I2**. Вчера обсудили, что формулировка **I2** была бы проще. Поэтому отдал ей приоритет (уступил в каком-то смысле).

Также попросил бы не "наезжать" при личном общении по телефону.

Author: Локального приложения не будет
I1 это то, что ты написал в доке.
Поправленное висит вверху.
Про предназначение или формулировку не понял.

Author: ```suggestion
Призначення системи. Порівняння з існуючими аналогічними системами. Ефект від її впровадження | <p>Основною задачею розробленої системи (тут і далі наз. **ОЗ**), у противагу до прикладної її складової, яка використовує та будує інфраструктуру навколо **ОЗ**), є визначення близькості текстових файлів (зокрема програмних кодів) у структурному та семантичному розумінні, візуалізація означеної близькості.</p><p>Інфраструктурна складова проекту використоує розроблену технологію **ОЗ** у задачі автоматичного оцінювання близькості (у означеному вище розумінні) текстових даних. Ми виділяємо дві інфраструктурні задачі, упорядковані за пріоритетністю виконання:<ul><li>автоматичне оцінювання близькості текстових файлів загальної форми (тут і далі **І1**). Сценарій користування наступний: <ul><li>Надати довільну кількість текстових файлів для подальшої обробки</li><li>Отримати діаграму близькості завантажених корпусів тексту за допомогою розробленого у рамках **ОЗ** рішення, </li><li>Детально вивчити різницю між двома підозрілими на схожість корпусами тексту через веб-інтерфейс</li></ul></li><li>автоматичне оцінювання близькості програмних кодів як задач з наявністю машинно-перевірюваного критерію правильності (тут і далі **І2**). Прикладом такої системи є тестовий сервер для перевірки програмного коду.</li></ul></p><p>**І1** та **І2** мають бути вирішені і розгорнуті у вигляді веб-сервера.</p></p>Авторам не відомий сервіс, що виконує задачі **ОЗ**, **І1** та **І2** разом. Не існує вирішення задач **І1**. Задача **І2** без використання **ОЗ** вирішується відомими сервісами Codeforces та ін. Не існує також розповсюдженого вирішення задачі **ОЗ**.</p><p>Задача **ОЗ** є дослідницьким напрямком, у якому state-of-the-art технології продовжуюсть з'являтися. Тому інтегрування таких технологій у прикладні рішення має цінність для стимулювання розвитку цієї галузі.</p>
```

Reviewer: Agree with this.

> Загальна характеристика задачі, що вирішується |<ul><li>Задачею семантичного представлення корпусу тексту займається дисципліна Natural Language Processing, що набула широкого розповсюдження у прикладних системах (наприклад, голосові помічники Amazon Alexa, Google Assistant). Пошук близькості між текстами може використовувати представлення у векторному просторі, як це роблять перелічені застосунки.</li><li>Невідомо, чи можемо ми модифікувати алгоритми пошуку структурної подібності для якісного вирішення задачі **ОЗ**. Частково мета даної роботи полягає в перевірці такого припущення.</li><li>Візуалізація близькості текстових корпусів - задача із багатьма можливими застосунками. </li><li>Проектування, розробка та розгортання веб-серверу, який виконував би задачу **ІЗ**, є необхідною частиною роботи для демонстрації успіху вирішення задачі **ОЗ**</li></ul>

Author: ```suggestion
Загальна характеристика задачі, що вирішується |<ul><li>Пошук близькості між текстами може використовувати представлення у векторному просторі, як це роблять застосунки, що вирішують задачі розпізнавання та обробки людської мови. (NLP)</li><li>Невідомо, чи можемо ми модифікувати алгоритми пошуку структурної подібності для якісного вирішення задачі **ОЗ**. Частково мета даної роботи полягає в перевірці такого припущення.</li><li>Візуалізація близькості текстових корпусів - задача із багатьма можливими застосунками. </li><li>Проектування, розробка та розгортання веб-серверу, який виконував би задачу **ІЗ**, є необхідною частиною роботи для демонстрації успіху вирішення задачі **ОЗ**</li></ul>
```

Author: Это не вода!

Author: Да поставь уже галочку, сколько можно. 21:52...

Author: В следующий раз не надо целый док заливать!

> Загальна характеристика задачі, що вирішується |<ul><li>Задачею семантичного представлення корпусу тексту займається дисципліна Natural Language Processing, що набула широкого розповсюдження у прикладних системах (наприклад, голосові помічники Amazon Alexa, Google Assistant). Пошук близькості між текстами може використовувати представлення у векторному просторі, як це роблять перелічені застосунки.</li><li>Невідомо, чи можемо ми модифікувати алгоритми пошуку структурної подібності для якісного вирішення задачі **ОЗ**. Частково мета даної роботи полягає в перевірці такого припущення.</li></ul>

Author: Обобщил детали, так как на данный момент ни ты, ни я, не обладаем досаточной информацией по поводу того, что конкретно мы из NLP и области машинного обучения мы будем использовать (PCA vs. T-sne, например). Такой подход позволит нам не ограничивать себя сейчас и избежать того, что мы "бросимся" в одно направление, и нам придется редактировать план из-за того, что поймем невыполнимость задачи. План не планирую редактировать на проятжении всего сместра, отсюда и желание сохранить "вписуемость" в рамки.

Reviewer: Заметь, я не писал, каким способом мы будем решать ту или иную задачу. Я лишь описал задачи, с которыми мы столкнёмся и какие есть способы решения для них.

    • Пошук близькості між двома текстовими файлами на даний момент має два основних підходи до рішення - TF-IDF, Word Embedding;
    • Приблизне або точне відображення точок з заданими попарно функціями відстані, що може бути виконане через теплову карту, деяку поверхню, а також інші алгоритми;
    • Пошук різниці між двома файлами через жадні diff алгоритми та інші алгоритми схожості;
    • Побудова архітектури веб-серверу з описаним вище функціоналом.

Reviewer: Ладно, соглашусь, что можно фереформулировать вот так.
• Пошук близькості між двома текстовими файлами;
• Приблизне або точне відображення точок з заданими попарно функціями відстані;
• Пошук різниці між двома файлами;
• Побудова архітектури веб-серверу з описаним вище функціоналом.

Author: > Пошук близькості між текстами може використовувати представлення у векторному просторі, як це роблять перелічені застосунки.
Тут имею в виду, что будут использоваться word embeddings.

- Разница между "Пошук близькості між двома текстовими файлами" и "Пошук різниці між двома файлами;"?

 • Побудова архітектури веб-серверу з описаним вище функціоналом.
Согласен.

• Приблизне або точне відображення точок з заданими попарно функціями відстані; Тут имеешь в виду проецирование точки в многомерном пространстве в дву(трех)мерное (t-sne, возможно, pca еще сюда относится).

Reviewer: Разница между "Пошук близькості між двома текстовими файлами" и "Пошук різниці між двома файлами;"?

NLP vs жадібні diff ?

> Имеешь в виду проецирование точки в многомерном пространстве в дву(трех)мерное (t-sne, возможно, pca еще сюда относится).
Не обязательно, я же не писал, что точки в многомерном пространстве - лишь описал, что между ними есть некая функция расстояния z = f(x,y) и нужно отобразить это в каком-то виде на график.

Author: Еще замечу. "Загальна характеристика задачі, що вирішується" и "Задача, що вирішується" не одно и то же. Здесь требуется пояснить, в чем сложность задачи / актуальность / новизна. Сама задача описана в предыдущем пункте. Иначе в чем смысл было бы дублировать одно и то же?

Тот факт, что есть семантика (NLP) и есть структурное сходство (diff-алгоритмы), а также что постановка задачи требует визуализацию определенного нами понятия "близости" (правда, опустил детали того, как мы это представляем математически) описан кратко выше:

> визначення близькості текстових файлів (зокрема програмних кодів) у структурному та семантичному розумінні, візуалізація означеної близькості.

Reviewer: В предыдущем пункте описано "призначення системи", а не задача, которая решается.

Author: Тем не менее, характеристика задачи и сама задача - не одно и то же.

Reviewer: Ладно, я на колл с великими людьми из Нидерландов.

Author: ```suggestion
Загальна характеристика задачі, що вирішується |<ul><li>Задачею семантичного представлення корпусу тексту займається дисципліна Natural Language Processing, що набула широкого розповсюдження у прикладних системах (наприклад, голосові помічники Amazon Alexa, Google Assistant). Пошук близькості між текстами може використовувати представлення у векторному просторі, як це роблять перелічені застосунки.</li><li>Невідомо, чи можемо ми модифікувати алгоритми пошуку структурної подібності для якісного вирішення задачі **ОЗ**. Частково мета даної роботи полягає в перевірці такого припущення.</li><li>Візуалізація близькості текстових корпусів - задача із багатьма можливими застосунками. </li><li>Проектування, розробка та розгортання веб-серверу, який виконував би задачу **ІЗ**, є необхідною частиною роботи для демонстрації успіху вирішення задачі **ОЗ**</li></ul>
```

> Мотивація вибору задачі |<ul><li>Дослідницька роботу у напрямку **ОЗ**.</li><li>Використання розв'язіків загально-теоретичних задач у прикладній області призведе до більшої зацікавленості до досліджень у напрямку методів розуміння семантики текстових даних.</li><li>Бажання вирішення проблеми справедливого оцінювання робіт шляхом візуалізації подібності рішень.</li></ul>

Author: Выбрал суть из предложенного ранее и сократил ее до минимума. Также убрал часть мотивации, которая касается "карьеры" и "чтобы разобраться", так как мне кажется, здесь не имеется в виду личная мотивация. Скорее мотивация выбора задачи исходя их объективных факторов.

Reviewer: Ну о каких обьективных факторах ты говоришь, если "Дослідницька роботу у напрямку **ОЗ**" не относится ко всем?????????????????????????????????//

Reviewer: Здесь имеется в виду именно личная мотивация выбора, так как другой мотивации быть не может.

Author: Часто когда пишут научные статьи, во введении поясняют актуальность исследования. Думал, здесь нужно аналогичное.

Reviewer: Да, там есть такой пункт, я же не спорю, но здесь его не спрашивали.
Можно и добавить, если есть желание.
Но вспомни структуру курсовой - этот пункт идёт как отдельный, где нужно обьяснить личную мотивацию по выбору задачи.

Author: > Здесь имеется в виду именно личная мотивация выбора, так как другой мотивации быть не может.

Может. Но верну на место личную мотивацию.

Author: > Ну о каких обьективных факторах ты говоришь, если "Дослідницька роботу у напрямку **ОЗ**" не относится ко всем?

Принимается. Подумал, это то же самое, что "здобування навичок та знань для роботи у прогресивному машинному навчанні".

Author: Передумал. Там бред. Я наоборот сократил ненужные вещи, сохранив суть. Что там добавлять. Нечего добавлять... Розпочати розвиток та карьеру в прогресивному напрямку машинного навчання. Цирк... Здесь не нужна личная мотивация. Личная мотивация получить оценку. Нет времени добавлять ненужные вещи обратно.

> Аналіз здійсненності. Необхідні для виконання проекту ресурси. Основний ризик для проекту | Складність полягає у плануванні розробки проекту, тобто рішенні підзадач зазначених у відповідних пуктах, проектуванні архітектури, опануванні фронт-енд частини. Ризиком є можлива недостатність часу для виконання бажаного обсягу роботи, що призведе до урізання функціоналу, вибрана ітеративна модель, що може сповільнити процес розробки, відсутність розробника, що на достатньому рівні вміє робити Front-End.

Author: Тут оставил как в доке. Не вижу изменений.

Reviewer: Согласен.
